### DISCUSSION TITLE
[N3DS] Swap Start/Select with ZL/ZR buttons

### DISCUSSION THREAD

**Discussion by Kyor-ob12** [diasurgical/DevilutionX#2869] [Ideas]
Having played a couple of hours and tested all characters, I came across to the difficulty of using hotkeys in my N2DS.

Taking a look to a picture of 3DS console family, all "New" labeled consoles have Start and Select buttons on the right side, making the use fo hotkeys for Spells (F1, F2, ...) a bit clunky.

Suggestion: 
- Switch Start <-> ZL and Select <-> ZR buttons
- Add a full example about how to compile code changing game controls, documentation seems to be uncomplete. Is ZR controlled by DKBCTRL_BUTTON_RIGHTSTICK? Which is its matching SDLK_* ?
https://github.com/diasurgical/devilutionX/blob/master/Source/controls/README.md

Great job! Keep improving it!

**Comment by StephenCWills** [comment]
This isn't a bad suggestion, but old 3DS doesn't have ZL/ZR so the game would become unplayable. I don't know how to reconcile that discrepancy without enabling button remapping after installation, which is a planned feature.

As for remapping the buttons at build time, the doc you linked is for emulating the controller using keyboard. The controls for 3DS are mapped in the n3ds_defs.cmake file in the repo so you can look there to see the variables and their values.

https://github.com/diasurgical/devilutionX/blob/9964bd9e495314e2d466eeb6a0c11d3c12c963ae/CMake/ctr/n3ds_defs.cmake#L39-L48

**Reply by Kyor-ob12** [reply]
Oh! I didn't realised about O3DS support, is nice to know about it!

Taking a look to that chunck of code, in order to perfor the button swap i want it should be like this:

` set(JOY_BUTTON_BACK 8) 
 set(JOY_BUTTON_START 9) 
 set(JOY_BUTTON_TRIGGERLEFT 7) 
 set(JOY_BUTTON_TRIGGERRIGHT 0) `

Thanks for your help to build using the correct values!

**Reply by Kyor-ob12** [reply]
I've just dowloaded the code and realised the existence of:
`sgOptions.Controller.bDpadHotkeys = GetIniBool("Controller", "Dpad Hotkeys", false);`

So it is presumed that putting this in your diablo.ini should set the Directional PAD as hotkeys:
```
[Controller]
Dpad Hotkeys=1
```
But has no effect in current release... May I recompile it by myself or I should add these values to devilutionX/CMake/ctr/n3ds_defs.cmake before doing so ?
```
  JOY_BUTTON_DPAD_LEFT
  JOY_BUTTON_DPAD_RIGHT
  JOY_BUTTON_DPAD_UP
  JOY_BUTTON_DPAD_DOWN
```

**Reply by StephenCWills** [reply]
I don't know of any reason why `Dpad Hotkeys` wouldn't work, recompiled or otherwise, although I haven't tried it myself. Can you clarify what you mean by "current release"? Maybe you should open an issue for it.

You shouldn't need the `JOY_BUTTON_DPAD_*` variables, unless you want to map some buttons as d-pad inputs. devkitPro updated their SDL binaries a few months ago with an update that models the d-pad as a hat instead of four individual buttons so that's what the `JOY_HAT_DPAD_*` variables are for.

**Reply by AJenbo** [reply]
`Dpad Hotkeys=1` was probably not part of 1.2.1, but only works in the latest test builds?

**Reply by StephenCWills** [reply]
I thought so too, but it seems to be there.
https://github.com/diasurgical/devilutionX/blob/1.2.1/SourceX/controls/game_controls.cpp#L161-L190
